### PR TITLE
Fix WebsocketController shutdown panic

### PR DESCRIPTION
**Pull request by TehTotalPwnage**
This PR fixes the panic that occurs when the `WebsocketController` begins shutting down its `Watchers` as described in rancher/longhorn#119. The issue was being caused because of how the slice of `Watchers` was being processed. Because of how the for loop was picking up the modifications of the slice, this would occur:

**Two asterisks means being currently processed in the for loop:**
```
// Processing index 0.
**watcher1** | watcher2 | watcher3 | watcher4
watcher2 | watcher3 | watcher4
// Now processing index1.
watcher2 | **watcher3** | watcher4
// watcher2 was skipped, now watcher3 is getting removed
watcher2 | watcher4
```
There's no third index here, but the for loop still had the value of watcher3 stored and attempted to close its channel twice, causing the issue.

There are a few other ways to rewrite the code to modify the slice properly. To leave in place most of the existing logic for how the slice gets processed, the slice now instead is iterated over backwards. That way, when a `Watcher` is removed from the end of the slice, it doesn't affect the ordering of the other elements inside, circumventing this issue.

This issue resolves rancher/longhorn#119.

### CODE REVIEW

**Comment by yasker** [comment]
The watcher shouldn't operate wc's list. It's websocket controller's responsibility.

We can move that part of the code to the wc. 

But is there any call to the `watcher.Close()` other than `wc.Close()`?

**Comment by TehTotalPwnage** [comment]
From what I've seen in the code, it looks like the WebsocketController is the only thing that's calling `watcher.Close()`.

**Comment by yasker** [comment]
@LLParse can you confirm that's the case?

**Comment by yasker** [comment]
@TehTotalPwnage Let's just do it for now. In the worst case, the channel will be GCed by the Golang.

**Comment by TehTotalPwnage** [comment]
Changes:
- The append function that removes the `Watcher` from the slice of `WebsocketController` has been removed. Removal of `Watchers` is now handled in the `Close` function of the `WebsocketController`.

File: controller/websocket_controller.go [hunk 1]
```diff
@@ -111,8 +104,14 @@ func (wc *WebsocketController) Run(stopCh <-chan struct{}) {
...
 
 func (wc *WebsocketController) Close() {
-	for _, w := range wc.watchers {
-		w.Close()
+	// Since the Close function on the Watcher modifies the slice in place, we
+	// could have problems if iterating over it in order (one item gets removed,
+	// next item gets pushed down to old index and skipped, etc.). To safely
+	// remove the items from the slice as we're doing now, we need to remove them
+	// from the end, which means we need to loop through our slice backwards.
+	for i := len(wc.watchers) - 1; i >= 0; i-- {
```
Comment (yasker) [inline controller/websocket_controller.go]: The lock is there to protect wc.watchers, not the single watcher.

You don't need to do it in reverse order now.

It should work if you just grab the lock, close everything in order in the loop, clean up the wc.watchers, and release the lock.

**Comment by TehTotalPwnage** [comment]
Changes:
- Lock management of the `Watchers` has also been relegated to the `Close()` function. The function will now just close all of the `Watchers` at once before clearing the slice and unlocking again.
- The only thing that `Watcher.Close()` now does is close that particular channel.

**Comment by LLParse** [comment]
Missed your ping, sorry bout that. LGTM

**Pull request merged**